### ISSUE TITLE
Add possibility to clear multiple cacheKey's or auto update cache.

### ISSUE THREAD

**Issue by luqezman** [joshuaslate/type-cacheable#211]
I started developing a simple CRUD api of TODO's and using type-cacheable.

The first problem encountered is "I cannot clear multiple cacheKey's". Why I want this? Because I cache a list of items using a specific cacheKey instead of id.

The second problem encountered is "I want to clear and update immediately", also for multiple cacheKey's. Why I want this? Because I want to clear the cache after updating and automatically re-cache after response.

I can open a PR to develop both of decorators. Is it possible?

Thanks

**Comment by luqezman** [comment]
My current behaviour:

```typescript

@Injectable()
export class TodoService implements ITodoService {
  constructor(private readonly todoRepository: TodoRepository) {}

  // TODO: After creation, TODO can be cached
  async createTodo(task: string): Promise<Todo> {
    const todoAsEntity = await this.todoRepository.create({ task });
    const todo = await this.todoRepository.save(todoAsEntity);

    return todo;
  }

  // I cannot cache a list of todos since I cannot delete that key because I prefer to cache a single TODO key
  async getAllTodos(): Promise<Todo[]> {
    const todos = await this.todoRepository.find();

    return todos;
  }

  @Cacheable({ cacheKey: ([id]: any[]) => id })
  async getTodoById(id: string): Promise<Todo | null> {
    const todo = await this.todoRepository.findOne(id);

    return todo;
  }

  @CacheClear({ cacheKey: ([id]: any[]) => id })
  async removeTodoById(id: string): Promise<boolean> {
    const { affected } = await this.todoRepository.delete(id);
    const hasBeenDeleted = affected > 0;

    if (hasBeenDeleted) {
      await cacheManager.client?.del('todos');
    }

    return hasBeenDeleted;
  }

  @CacheClear({ cacheKey: ([{ id }]: any[]) => id })
  async updateTodo(todo: Todo, dataToUpdate: TodoUpdateInput): Promise<Todo> {
    const todoToUpdate = Object.assign(todo, dataToUpdate);
    const updatedTodo = await this.todoRepository.save(todoToUpdate);

    await cacheManager.client?.del('todos');

    return updatedTodo;
  }
}
```

Like I said before I want to clear other keys because I want to cache a list of todos in my method `getAllTodos`.

So I suggest caching results immediately and adding a possibility to clear another keys:

```typescript
  @CacheResult({ cacheKey: (todo) => todo.id, keysToClear: 'todos' /* Possibility to clear another keys */ })
  async createTodo(task: string): Promise<Todo> {
    const todoAsEntity = await this.todoRepository.create({ task });
    const todo = await this.todoRepository.save(todoAsEntity);

    return todo;
  }

  @CacheClear({ cacheKey: ([id]: any[]) => [id, 'todos'] /* Possibility to clear multiple keys */ })
  async removeTodoById(id: string): Promise<boolean> {
    const { affected } = await this.todoRepository.delete(id);
    const hasBeenDeleted = affected > 0;

    if (hasBeenDeleted) {
      await cacheManager.client?.del('todos');
    }

    return hasBeenDeleted;
  }

  @CacheResult({ cacheKey: (todo) => todo.id, keysToClear: 'todos' /* Possibility to clear another keys */ })
  async updateTodo(todo: Todo, dataToUpdate: TodoUpdateInput): Promise<Todo> {
    const todoToUpdate = Object.assign(todo, dataToUpdate);
    const updatedTodo = await this.todoRepository.save(todoToUpdate);

    await cacheManager.client?.del('todos');

    return updatedTodo;
  }
```

Thanks again

**Comment by joshuaslate** [comment]
#212 will add the ability to pass multiple keys to clear for the next release. I'll work on the rest in the (hopefully) near future, else I'm happy to review PRs.

**Comment by luqezman** [comment]
@joshuaslate wow! You were quite quick!

What about @CacheResult idea,cCaching immediately after the result? It prevents concurrent requests trying to build the cache simultaneously.

**Comment by joshuaslate** [comment]
I agree we should support that, I'm just trying to find the best way. I originally thought composition might work here, e.g.:

```
@Cacheable({ cacheKey: (todo) => todo.id })
@CacheClear({ cacheKey: (todo) => todo.id })
async updateTodo(todo: Todo, dataToUpdate: TodoUpdateInput): Promise<Todo> {
    const todoToUpdate = Object.assign(todo, dataToUpdate);
    const updatedTodo = await this.todoRepository.save(todoToUpdate);

    return updatedTodo;
  }
```

But had no luck with this, unfortunately.

Maybe something like `@CacheUpdate` or `@CacheReplace`? I think it's more explicit than `@CacheResult`, but the concept would be the same as yours above.

**Comment by luqezman** [comment]
@joshuaslate when the new release will be available?

**Comment by luqezman** [comment]
@joshuaslate I prefer @CacheUpdate, it's more readable. But I think you cant use @Cacheable in a persist method since you dont want to cache that method, I just want to replace immediately a single cache key and clearing another keys. In this case I want to "update" current todo and clear the key 'todos' which is responsible for caching a list, in my example.

**Comment by joshuaslate** [comment]
v6.1.0 is published now with the ability to provide `@CacheClear` with an array (or function that returns an array) for the `cacheKey` value.

As for the `@CacheUpdate` decorator, sure. I can do this, but I'll need to look at it another day.

**Comment by luqezman** [comment]
Big thanks @joshuaslate <3

**Comment by luqezman** [comment]
@joshuaslate mark this issue as resolved/merged, etc. Thanks.

**Comment by joshuaslate** [comment]
`@CacheUpdate` has landed in 6.2.0, please let me know if it suits your purposes. It's working for me.

**Comment by luqezman** [comment]
Yes, everything is working here! Big hug <3
